services/npm/npm-last-update.service.js

```diff
@@ -3,55 +3,111 @@ import dayjs from 'dayjs'
 import { InvalidResponse, NotFound, pathParam, queryParam } from '../index.js'
 import { formatDate } from '../text-formatters.js'
 import { age as ageColor } from '../color-formatters.js'
-import NpmBase, { packageNameDescription } from './npm-base.js'
+import NpmBase, {
+  packageNameDescription,
+  queryParamSchema,
+} from './npm-base.js'
 
-const updateResponseSchema = Joi.object({
-  time: Joi.object({
-    created: Joi.string().required(),
-    modified: Joi.string().required(),
-  })
+const fullSchema = Joi.object({
+  time: Joi.object()
     .pattern(Joi.string().required(), Joi.string().required())
     .required(),
   'dist-tags': Joi.object()
     .pattern(Joi.string().required(), Joi.string().required())
     .required(),
 }).required()
 
-export class NpmLastUpdate extends NpmBase {
+const abbreviatedSchema = Joi.object({
+  modified: Joi.string().required(),
+}).required()
+
+class NpmLastUpdateBase extends NpmBase {
   static category = 'activity'
 
-  static route = this.buildRoute('npm/last-update', { withTag: true })
+  static defaultBadgeData = { label: 'last updated' }
+
+  static render({ date }) {
+    return {
+      message: formatDate(date),
+      color: ageColor(date),
+    }
+  }
+}
+
+export class NpmLastUpdateWithTag extends NpmLastUpdateBase {
+  static route = {
+    base: 'npm/last-update',
+    pattern: ':scope(@[^/]+)?/:packageName/:tag',
```

Comment:
Given how different the two implementations for with/without tag are, I've just split this into two classes. One with a (required) tag param and one without it. This means we need to specify the routes manually instead of using the `buildRoute()` helper, which makes the `tag` param optional.

Reply:
Not sure if it improves anything but we could use `buildRoute('npm/last-update' ,{ withTag = false })` here and keep the required tagged route as is.